[vcpkg] Rewriting CmdLineBuilder/Command (3/n)

strega-nil opened the pull request:
Rename CmdLineBuilder to Command, since it's no longer a builder but an actual data type.

This is the 3rd PR in the "rewrite CmdLineBuilder" saga; at this point, we need a new type name :)

BillyONeal commented on toolsrc/include/vcpkg/base/system.process.h:
```diff
@@ -20,26 +20,26 @@ namespace vcpkg::System
     };
 
-    struct CmdLineBuilder
+    struct Command
```
I don't like this name; it is too generic since the command pattern is a fairly general concept that we might want to use in the future.

How about `CommandLine` or `SubprocessInfo` or `LaunchInfo` or something like that?

strega-nil replied:
Fair enough, it's named after the Rust type.

BillyONeal commented on toolsrc/include/vcpkg/base/system.process.h:
```diff
@@ -51,17 +51,17 @@ namespace vcpkg::System
     };
 
-    struct CmdLineBuilderMapLess
+    struct CommandMapLess
```
Should be `CommandLineLess` -- that we use it in a map isn't intrinsic to what it does.

strega-nil replied:
This is the standard way we call this; `XMapLess` means "a `Less` implementation that is not to be used for ordering, just for use in maps".

BillyONeal replied:
But it's totally reasonable to order these by resulting command line outside of the context of a map.

I would understand if this comparison was totally useless otherwise but that's not the case here.

strega-nil committed:
[vcpkg] Rewriting CmdLineBuilder/Command (3/n)

Rename CmdLineBuilder to Command, since it's no longer a builder but an actual data type

ras0219 commented on toolsrc/include/vcpkg/base/system.process.h:
```diff
@@ -20,26 +20,26 @@ namespace vcpkg::System
+        explicit Command(const std::string& s) { string_arg(s); }
+        explicit Command(const char* s) { string_arg({s, ::strlen(s)}); }
+
+        Command& path_arg(const fs::path& p) & { return string_arg(fs::u8string(p)); }
```
I'm a bit unhappy with the combination of the "data" name plus "data" named functions; specifically, it is not clear from the name of the type nor the name of the functions that these are mutating functions -- you need to either notice that the return type is `&` or read the implementation to know that these will mutate the existing object.

This is why I named the original type "Builder": so that the functions like `path_arg` could be short instead of `append_path_arg()` yet there was still nomenclature to note that these functions are mutating.

ras0219 replied:
I also generally don't see sufficient motivation for change in the name, this just introduces churn imo.
